distribution/xfer/download.go

```diff
@@ -140,11 +178,14 @@ func (ldm *LayerDownloadManager) Download(ctx context.Context, initialRootFS ima
 		progress.Update(progressOutput, descriptor.ID(), "Pulling fs layer")
 
 		var xferFunc DoFunc
+		cacheKey := descriptor.ID()
```

Comment:
I don't think `descriptor.ID()` is a good cache key, because it's a truncated digest, so it could collide with another download.

Even if we used non-truncated digests, I think there are still cases where we could be downloading the same layer more than once at the same time. Downloads in the download manager are deduplicated with a key composed of the entire stack of layers so far (`transferKey`), not simply the blob digest. IIRC this was done on purpose to avoid the difficulty of reference counting download files. In the current design, each download will only be registered into the layer store once. In the rare case that multiple concurrent pulls reference the same layer on top of a different parent layer, that layer will be downloaded multiple times.

So for now I would recommend using `transferKey` as the cache key to avoid corner cases where multiple downloads could try to use the same temporary file at once. In a followup, we could potentially do further refactoring so that we separate the notion of downloading a layer from registering it, and allow deduplication of downloads even when they would be registered on top of different parents (however, I think the benefit would be relatively low).